src/config.c

```diff
@@ -1678,6 +1678,8 @@ int rewriteConfigOverwriteFile(char *configfile, sds content) {
         serverLog(LL_WARNING, "Could not chmod config file (%s)", strerror(errno));
     else if (rename(tmp_conffile, configfile) == -1)
         serverLog(LL_WARNING, "Could not rename tmp config file (%s)", strerror(errno));
+    else if (fsyncFileDir(configfile) == -1)
```

Comment:
I agree
I didn't think over the failure case :( 
maybe we could cleanup with `if (retval && fileExist(tmp_conffile)) unlink(tmp_conffile)` which not break the previous style

Reply:
I don't think there is a simple solution for handling directory fsync() failures. Ideally, if directory fsync() fails, it should rollback rename(). I'm not sure if we can do it without a rollback journal. If we are not rolling it back, probably there is no point in error handling. I'd just ignore directory fsync() error. Also, IIRC there are a few filesystems, probably not the mainstream ones, they don't support fsync on a directory. I remember fsync() on a directory would return an error on these filesystems.

Reply:
@yossigo pointed me to this thing: https://danluu.com/fsyncgate/ indicating that (at last on some Linux versions), data that failed to sync on one fsync call, will never get synced, even if a consecutive call succeeds.
i'm not sure if it applies for directory fsync too, or if it's still valid for recent Linux versions.

maybe that's indicating that we should always panic in that state, or alternatively become read-only like we do on persistence errors, since returning an error to the user is not always possible, but considering what i wrote above, it may not be enough.

I'd like to know what PostgreSQL and SQLite are doing.. maybe mimicking them is sufficient.

Reply:
I have little knowledge about PostgreSQL or SQLite, it's a too complex domain to dive in.. hope there're someone else good at it.

But I'm familiar with Leveldb, and have some glances at Rocksdb. They're key-value storage library, providing APIs to uplevel applications. Many famous database like MyRocks, TiDB and Ceph BlueStore is based on Rocksdb. As I mentioned before, they would just expose the IO errors(including fsync file or dir) to the uplevel caller, and the deal completely depends on the uplevel application. I try to search keyword `EIO` which mentioned in the article but nothing special is found.

Refer to it, my idea is dividing the errors into two classes to handle, ones triggered by user command like `config rewrite` and `bgsave`, and other triggered by background persistence like RDB or AOF routine. For the former case,  Redis just notify user the error,  it's up to users to retry, debug or failover, and the failure wouldn't result in data loss or corruption, as we use an atomic "create new temp, write and rename" pattern. For the latter case, Redis would turn into read-only, as more write is at a risk of data loss mentioned in the article.

Reply:
I tried to look at these dbs.

1- SQLite ignores fsync error on the directory. You can see it in the link above.
2- PostgreSQL: I'm not sure. There are cases it ignores, there are cases just prints a log and there are cases that exists the process with an error code (Not the main process, process that is created for a task). 

In any case, I think we should borrow the checks here to detect filesystems that do not support fsync on a directory:

https://github.com/postgres/postgres/blob/55b56865115eccd6449e79d6f06fe49d6ba3b792/src/backend/storage/file/fd.c#L3708

In summary: 
- If `open()` returns `EISDIR` or `EACCES` (`EACCES` might be Windows only but we can keep it, just in case.)
- If `fsync()` returns `EBADF` or `EINVAL`

means it's a system that does not support either `open()` or `fsync()` on a directory. We should just skip syncing the directory in this case.

Reply:
@tezc thank you, following your advise I ignore some error for `open()` and `fsync()`. RocksDB and PostgreSQL seem to take a different way to deal with the cross-platform issues. RocksDB have some special handling for some OSs with condition compilation, in contrast PostgreSQL ignore some specific errno safely. For safety I adopt both of the two way.